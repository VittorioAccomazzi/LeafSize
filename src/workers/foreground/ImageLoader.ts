import {noExtension} from '../../common/utils/FileUtils'

type Quadrant = 'Q0' | 'Q1' | 'Q2' | 'Q3' | 'All'

// this interfce is designed to facilitate the testing (inversion of dependencies )
export interface ImgLoaderFileHandle {
    name : string,
    getFile : ()=>Promise<ImgLoaderFile>
}


export interface ImgLoaderFile {
    arrayBuffer : () => Promise<ArrayBuffer>
}

interface ImageTile {
    file :  ImgLoaderFileHandle //FileSystemHandle,
    name : string,
    quadrant : Quadrant
}

export type ImageFileLoader =  ( handle : ImgLoaderFileHandle ) => Promise<ImageBitmap> 

/**
 * Image loader.
 * Load the images from file and split them up as necessary.
 */
export default class ImageLoader  {
    private tiles : ImageTile[] = []
    private cacheImage : ImageBitmap | null = null;
    private cacheFile  : ImgLoaderFileHandle | null = null;
    private getImageFromFile : ImageFileLoader
    private targetSize : number;

    /**
     * construct a ImageLoader object
     * @param files list of files to load
     * @param numDishes  number of dishes per image : 1 or 4
     * @param targetSize  target size for the image's diagonal.
     * @param imageFileLoader loader function (if required)
     */
    constructor ( files : ImgLoaderFileHandle[] | null , numDishes : number, targetSize : number, imageFileLoader : ImageFileLoader = getImageFromFile ){
        this.tiles = []
        this.getImageFromFile = imageFileLoader;
        this.targetSize = targetSize;
        if (numDishes !==1 && numDishes !== 4 ) throw new Error(`not supported num Dishes ${numDishes}`)

        if( files ){
            switch(numDishes){
                case 1: {
                    this.tiles = files.map(file => ({ file, quadrant : 'All', name : noExtension(file.name)}))
                    break;
                }
                case 4: {
                    files.forEach(file =>{
                        ['Q0','Q1','Q2','Q3'].forEach(quadrant=>{
                            this.tiles.push({
                                file,
                                name : noExtension(file.name)+'-'+quadrant,
                                quadrant : quadrant as Quadrant
                            })
                        })
                    })
                    break;
                }
            }
        }
    }

    /**
     * list of images generated
     */
    get List() : string [] {
        return this.tiles.map(f=>f.name);
    }

    /**
     * get file handle given the image name
     */
    FileHandle( name : string ) :ImgLoaderFileHandle | undefined {
        return this.tiles.find(v=>v.name===name)?.file;
    }

    /**
     * number of dishes for image
     */
    get NumDishes() : number {
        let qList = new Set(this.tiles.map(v=>v.quadrant));
        return qList.size;
    }

    /**
     * Extract an image from the file.
     * @param inCanvas canvas where the image will be copied to.
     * @param index index of the image to extract
     * @returns scale : downsample value of the original image in order to reach the diagonal size set in the constructor. imageData : image Data for the canvas object
     */
    async getImage( index : number ): Promise< { scale : number, imgData : ImageData|null }> {
        let scale = 0;
        let imgData = null;
        if( !this.tiles || !this.tiles.length ) return {scale, imgData }; // this can happen as transition
        if( index <0 || index >this.tiles.length) throw Error(`undexpected index : requested ${index} but got ${this.tiles.length}`)
        const tile = this.tiles[index];
        const img = await this.getImageFromCache( tile );
        const imgWidth = img.width;
        const imgHeight= img.height;
        let xTopLft = 0;
        let yTopLft = 0;
        let xBtmRgt = imgWidth;
        let yBtmRgt = imgHeight;
        const xMid = imgWidth/2 | 0;
        const yMid = imgHeight/2 | 0;
        switch( tile.quadrant ){
            case 'Q0' : {
                xTopLft = 0;
                yTopLft = 0;                
                xBtmRgt = xMid;
                yBtmRgt = yMid;
                break;
            }
            case 'Q1' : {
                xTopLft = xMid;
                yTopLft = 0;
                xBtmRgt = imgWidth;
                yBtmRgt = yMid;
                break;
            }
            case 'Q2' : {
                xTopLft = 0;
                yTopLft = yMid;                
                xBtmRgt = xMid;
                yBtmRgt = imgHeight;
                break;
            }
            case 'Q3' : {
                xTopLft = xMid;
                yTopLft = yMid;                
                xBtmRgt = imgWidth;
                yBtmRgt = imgHeight;
                break;
            }
        }
        const srcWidth = xBtmRgt-xTopLft;
        const srcHeight= yBtmRgt-yTopLft;
        const srcDiag = Math.sqrt(srcHeight*srcHeight+srcWidth*srcWidth);
        scale = Math.min( 1, this.targetSize/srcDiag);
        const dstWidth = srcWidth * scale;
        const dstHeight= srcHeight * scale;

        // we need to get to the pixel data, and so we need a canvas (unfortunately)
        // see https://github.com/whatwg/html/issues/4785
        // Note OffscreenCanvas is deprecated below because it is experimental 
        // see https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas/OffscreenCanvas 
        // however since this will run in the workers I have no othre option.
        let canvas : HTMLCanvasElement | OffscreenCanvas;
        if( typeof document === 'undefined' ) {
            // worker
            canvas = new OffscreenCanvas(dstWidth, dstHeight);
        } else {
            // window
            canvas = document.createElement('canvas')
            canvas.width = dstWidth;
            canvas.height= dstHeight;
        }

        // Extract the pixel data which is needed for further processing.
        const ctx = canvas.getContext('2d');
        if( ctx ) { 
            ctx.drawImage( img, xTopLft, yTopLft, srcWidth, srcHeight, 0, 0, dstWidth, dstHeight);
            imgData = ctx.getImageData(0,0,dstWidth, dstHeight);
            // Debug only
            // ctx.font = '48px serif'; // debug only.
            // ctx.fillText(tile.name, 10, 60);
        }
        return { scale, imgData }
    }

    /**
     * This function perform some simple caching on the images because when the user is settings 4 dishes per image,
     * we can effectively load the same image for 4 different request.
     * @param tile tile which describes the images to extract.
     * @returns 
     */
    private async getImageFromCache( tile : ImageTile  ) : Promise<ImageBitmap> {
        if( this.cacheFile !== tile.file  ){
            if( this.cacheImage ) this.cacheImage.close(); // free resources.
            this.cacheImage = await this.getImageFromFile( tile.file );
            this.cacheFile  = tile.file;
        } 
        return this.cacheImage!;
    }
}

    /**
     * Load the image from file.
     * @param handle file to load
     * @returns 
     */
     async function getImageFromFile( handle : ImgLoaderFileHandle ) : Promise<ImageBitmap> {
        const file = await handle.getFile();
        const data = await file.arrayBuffer();
        const blob = new Blob([data]);
        return createImageBitmap(blob);
    }