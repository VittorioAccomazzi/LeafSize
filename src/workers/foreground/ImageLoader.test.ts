import ImageLoader , {ImgLoaderFileHandle, ImgLoaderFile} from './ImageLoader';
import {loadImage } from 'canvas';
import hash, { loadCanvas, testImage } from '../../common/utils/TestUtils'
import CanvasUtils from '../../common/utils/CanvasUtils';


// for the interface declaration with constructor refer to 
// https://fettblog.eu/typescript-interface-constructor-pattern/

class FileHandleMock<T extends ImgLoaderFile> implements ImgLoaderFileHandle {
    private filename = '';
    private factory : new ()=> T;

    constructor ( name : string, factory : new ()=>T ){
        this.filename = name
        this.factory = factory;
    }
    get name() : string {
        return this.filename;
    }
    async getFile () {
        const val : T = new this.factory();
        return val;
    }
}

describe('ImageLoader File list',()=>{

    class EmptyFileMock implements ImgLoaderFile  {
        constructor() {}
        async arrayBuffer() {
            return new ArrayBuffer(0)
        }
    }

    const files : FileHandleMock<EmptyFileMock> [] =[
        new FileHandleMock<EmptyFileMock>('image1.jpg',EmptyFileMock),
        new FileHandleMock<EmptyFileMock>('file.txt',EmptyFileMock),
        new FileHandleMock<EmptyFileMock>('name',EmptyFileMock)
    ]

    const names1 = [
        'image1',
        'file',
        'name'
    ]

    const names4 = [
        'image1-Q0',
        'image1-Q1',
        'image1-Q2',
        'image1-Q3',
        'file-Q0',
        'file-Q1',
        'file-Q2',
        'file-Q3',
        'name-Q0',
        'name-Q1',
        'name-Q2',
        'name-Q3'
    ]
    
    test('dish 1 use case',()=>{
        const imageLoader = new ImageLoader(files, 1, 1000 );
        expect(imageLoader.List).toStrictEqual(names1)
    })

    test('dish 4 use case',()=>{
        const imageLoader = new ImageLoader(files, 4, 1000);
        expect(imageLoader.List).toStrictEqual(names4)
    })

    test('shall throw if dishes is wrong',()=>{
        expect(()=>{
            new ImageLoader(files, 3, 1000)
        }).toThrow()
    })
      
})


describe('ImageLoader', ()=>{

    const imageDownsample = 4;

    class CanvasFileMock implements ImgLoaderFile  {
        constructor() {}
        async arrayBuffer() {
            const canvas =  await loadCanvas(testImage,imageDownsample);
            return canvas.toBuffer().buffer;
        }
    }

    const files : FileHandleMock<CanvasFileMock> [] =[
        new FileHandleMock<CanvasFileMock>('SampleImage.jpg',CanvasFileMock)
    ];


    const getImageFromFile = async ( handle : ImgLoaderFileHandle) : Promise<HTMLImageElement> => {
        const file = await handle.getFile();
        const buffer= await file.arrayBuffer();
        const image = await loadImage(Buffer.from(buffer));
        return image as unknown as HTMLImageElement; 
    }

    it('shall split the images properly', async ()=>{

        let numInvocations = 0;
        const getImageWithCount= async ( handle : ImgLoaderFileHandle) : Promise<HTMLImageElement> => {
            numInvocations++
            return getImageFromFile(handle);
        }

        const imgLoader = new ImageLoader( files, 4, 600, getImageWithCount );
        expect(imgLoader.List.length).toBe(4);
        const canvas = document.createElement('canvas');
        for( let i=0; i<imgLoader.List.length; i++ ){
            const {imageData} = await imgLoader.getImage(i);
            CanvasUtils.PutImageData(canvas,imageData!);
            const fileName = `canvas-${i}`;
            const hsh = await hash(canvas, fileName);
            expect(hsh).toMatchSnapshot();
        }

        expect(numInvocations).toBe(1);

    })

    it('shall scale the image properly', async ()=>{
        // this test relies on the actual image size
        const orgWidth = 6000;
        const orgHeight= 4000;
        let actWidth = orgWidth/imageDownsample;
        let actHeight= orgHeight/imageDownsample;
        actWidth /= 2; // since we have 4 dishes in an image
        actHeight/= 2; 
        const actualDig= Math.sqrt(actWidth*actWidth+actHeight*actHeight);
        const expScale = 0.5;
        const expectDig= actualDig*expScale;

        const imgLoader = new ImageLoader( files, 4, expectDig, getImageFromFile );
        expect(imgLoader.List.length).toBe(4);
        const canvas = document.createElement('canvas');
        for( let i=0; i<imgLoader.List.length; i++ ){
            const  { scale, imageData} = await imgLoader.getImage(i);
            CanvasUtils.PutImageData(canvas, imageData!);
            expect(scale).toBeCloseTo(expScale,2);

        }
    })

})