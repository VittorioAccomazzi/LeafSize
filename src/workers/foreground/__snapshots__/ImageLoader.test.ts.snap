// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoader shall split the images properly 1`] = `"fffffffffffffffffff4000000000000ffffffffffffffffffc0000000000000fffffffffffffffffe00000000000000ffffffffffffffffff00000000000000fffffffffffffffff800000000000000ffffffffffffffff4000000000000000fffffffffffffff00000000000000000ffffffffffffffc00000000000000000ffffffffffffff000000000000000000fffffffffffffc000000000000000000fffffffffffff0000000000000000000ffffffffffffc4004000000000000000ffffffffffff901ffe00000000000000fffffffffffe60ffffc0000000000000fffffffffffc83fffff4000000000000fffffffffffb07fffffc000000000000fffffffffff61fffffff000000000000ffffffffffec3fffffff400000000000ffffffffffd87fffffffc00000000000ffffffffffb0ffffffff800000000000ffffffffff61fffffff9800000000000fffffffffee3fffffff8000000000000fffffffffdc7ffffff50000000000000fffffffffd87ffffff00000000000000fffffffffb07fffffe00000000000000fffffffff70f7ffff600000000000000fffffffff61f7abff000000000000000ffffffffee3e001ff000000000000000fffffffffc7e000fb000000000000000ffffffffd97c0007b000000000000000fffffffff8fc00018000000000000000ffffffffb3fe00008000000000000000ffffffffffff0003fffe00000c3ffff4fffffffffffe0001fffe0000043fffb6fffffffffffe0000fffe0000061ffdfefffffffffffc00007ffc0000021ffff8fffffffffffc00003ffc0000020fffe1fffffffffff800003fd00000010ffff0fffffffffff000003fc00000000fffdcfffffffffff000003fc000000007fff0ffffffffffe000007f8000000007fff0ffffffffffe000007f8000000007fff4ffffffffffc000007f0000000047fff4ffffffffffc000007f0000000043ffe3ffffffffff8000007e0000000043ffcaffffffffff800000300000000063ffe7ffffffffffc00000200000000061ffa6ffffffffff800000000000000071ffeeffffffffff800000007800000031ff9cffffffffff000000007800000031ff0cfffffffffe00000000f800000031fe1dfffffffffc00000000f800000030ffbefffffffffc00000000f800000010fffefffffffffc00000000f000000010fffdfffffffffc00000001f000000010fffeffffffffffc0000001e000000010ffffffffffffff80000001e000000010ffffffffffffff80000003c0000000107ff4ffffffffff00000007c0000000007ff2ffffffffff00000007c0000000087fedffffffffff0000000fc0000000187fc8fffffffffe0000000f80000002987ff8fffffffffe0000000f80000002887ff8fffffffffe0000001f000000069c7ff8fffffffffc0000001f00000007983fd8fffffffffc0000001f00000007883ff2fffffffffc0000001f00000006c87fe8ffffffffff0000001e00000006c87fe4ffffffffff0000001e00000006887fe0ffffffffff0000001e00000004887fb1ffffffffff0000001f80000004887f3dffffffffff0000001fc00000048877d0ffffffffff8000001fe0000004883fb8ffffffffff8000003ff0000004887ff1ffffffffff8000007ff0000004983ef0ffffffffff800000fff8000000987ef8ffffffffff800001fff8000000903f88ffffffffff800003fff0000000907fc4ffffffffff800003fff0000000107e40ffffffffff800007fff80000001076c0ffffffffff800007fff8000001107fc0ffffffffff800007fff8000001107e60ffffffffff800007fffc000000005c00ffffffffff80000ffffc00000000fcc0ffffffffff80000ffff800000020fd40ffffffffff80000ffff800000020ed80ffffffffff80001ffff8000000003560ffffffffff80001ffff8000000003640ffffffffffc0001ffff40000000036c4ffffffffffc0001ffff4000000005c60ffffffffffc0003ffff2000000007608ffffffffffc0003ffff2000000012210ffffffffffc0007ffff8000000819380ffffffffffc0007ffff800000003db00ffffffffffc0007ffff8000000037a00ffffffffffe000fffff0000000007c00ffffffffffe0007fff40000000000000ffffffffffe0007fff40000000000000ffffffffffe0007fffc0000000000000ffffffffffe000ffffa0000000000000ffffffffffe000ffff80000000000000ffffffffffe000ffff80000000000000ffffffffffc080ffffc0000000000000ffffffffffc0c1ffffc0000000000000ffffffffffe0c1ffffe0000000000000fffffffffff1c1ffffe0000000000000fffffffffffbc1fffff0000000000000ffffffffffffe1fffff0000000000000fffffffffffff1fffff0000000000000ffffffffffff73ffff70000000000000ffffffffffb30bffff10000000000000ffffffffffc00bffff00080000000000ffffffffffe009ffff80080000000000fffffffffff001ffff84100000000000fffffffffff800ffff2fc00000000000fffffffffffc00bffb0e800000000000fffffffffffe003dd904000000000000ffffffffffff0000c018000000000000ffffffffffffc0000000000000000000ffffffffffffe0000000000000000000fffffffffffff8000000000000000000ffffffffffffff000000000000000000ffffffffffffffe00000000000000000fffffffffffffffff000000000000000fffffffffffffffffff0000000000000ffffffffffffffffffff000000000000ffffffffffffffffffff800000000000ffffffffffffffffffff000000000000"`;

exports[`ImageLoader shall split the images properly 2`] = `"fffdff0000003ffffe00bfffffff004ffffe80000000801c00007fffffff004ffffeb2000000000000001fffffff004fffffb9000040000000001fffffff004fffffff800000000000001fffffff004ffffff60000000000000003ffffff00cfffffde0000000000000001ffffff00cffffffa0800000000000000bfffff00dffffffe300004000000001f3fffff00dffffffd00000c0003f8003fffffff00dffffffc430030000fff001fffffff00dfffffff0b00600067fe000fffffff005fffffffdb00c00014060007ffffff005fffffeffa038000400f8003ffffff005fffffffbc03000002321001ffbfff005ffffffffc0c00000c007c00ff9fff007ffffffff01c000060049f007fffff03fffffffff0380000001c0a003fffff07ffffffffe070001008205f801fffff07ffffffffe0f0000000e4ff801fffff07ffffffffc1e00000090f0ff00fffff07ffffffffc1c0000000361ff007ffff07ffffffff8180000000040ff007ffff07ffffffff07800080000c0bf803ffff0cffffffff0701000000000ffe01ffff0c7fffffff0e020180000002fe05ffff00ffffffff1e0001000000003e00ffff00fffffffe1c0000000000001f82ffff0cfffffffc1c0000000000001f007fff0ffffffffc1c0800000000000f817fff0ffffffff8381000000000001f007fff0ffffffff8381000000000000fc03fff0ffffffffdfc3000000000001ff1bfff0ffffffffff82000000000001ff9dfff0ffffffffff06000000000000ff8dfff0ffffffffdf04000000000000ffcffff0ffffffffff0c0000000000007fcffff0cffffffffe0c0000000000007feffff0cffffffffe080000000000003feffff00fffffffbe180000000000007ff7fff04ffffffffc180000000000003ff7fff0dffffffefc100000000000003ff7fff0ffffffff7c700000000000003ffffff1ffffffff7c700000000000003ffffff1fffffffffce00000000000003ffffff0fffffffff8e00000000000001ffffff0ffffffff7ce00000000000001ffffff0fffffffffcc00000000000001ffffff1fffffffff8c00000000000001ffffff1dffffffff9c00000000000001ffffff1cffffffef9c00000000000001ffffff1dffffffef9c00000000000001ffffff00ffffffff9c00000000000001ffffff1dffffffff1800000000000003ffffff1fffffffff1800000000000001ffffff1fffffffff1800000000000001ffffff1fffffffff1800000000000001ffffff1fffffffff3800000000000000ffffff0fffffffff1800000000000000ffffff1fffffffff1800000000000000ffffff1fffffffff18000000000000007fffff1effffffff18000000000000007ffff9c8ffffffff08000000000000003ffff9fdffffffff08000000000000003ffff9ffffffffff88000000000000003ffff9ffffffffff8c000000000000003ffff9ffffffffff8c000000000000003ffff9ffffffffff0e00000000000000fffff9ffffffffef0600000000000001fffff9f9ffffffff0600000000000001fffffcf9ffffffef0600000000000001fffffcc0ffffffef0700000000000001fffffcc0f3ffffef0700000000000001fffffcfbe0ffffff0700000000000001fffffce1e8ffffff0700000000000001fffffce0e8ffffff0300000000000003fffffce0e8ffffff0380000000000003fffffce1e8fffff703c0000000000003fffffcffe0fffff70380000000000003fffffcffe1fffff70380000000000003fffffcfffffffffb8380000000000003fffffcfffffffffb8380000000000003fffffcfffffffffb81c0000000000003fffffc7ffffffffb8180000000000003fffffc7ffffffffd8180000000000003fffffc7ffffffffd80c0000000000003fffffc7ffffffff9c0e0000000000003fffffc7fffffffffc080000000000003fffffc7fffffffffc000000000000003fffffe7ffffffffe8000000000000003ffffff7fffffffff4000000000000007ffffff1fffffffff0000000000200007ffffff1fffffffffa000000000f00007ffffff1fffffffffa000000000f80007ffffff1fffffffff9000000003f80007ffffff0fffffffffd000000001fc000fffffff1fffa000000000000000080003caffff1fff00000000000000001c000399ffff1fff1c000000000000001800031dffff1cff00000000000000000e00033bffff1cffe0000000000000007a00072bffff00ffc000000000000000fe000617ffff1cffc800000000000000ff00000fffff1fff08000000000000007bc0000fffff1fff00000000000000003fe0001fffff1fffc0000000000000007fc0013fffff1fff00000000002000003ff0013fffff0fff00000000001000007ff8007fffff1fff000000020000000007f800ffffff1fff4000000200000000149001ffffff1fff0000000080000000004003ffffff1cff0000000000000000000007ffffff0cff700000043400000000000fffffff03ff880000059c00000000001fffffff0cff820000057700000000007fffffff1fff80000004278000000000ffffffff1fff0000000007e000000003ffffffff0fffdfe41000047800000007ffffffff0ffffaf7f7fffffc0000003fffffffff0fffff7f7793ffffffc001ffffffffff0ffffb6bf7d89fffffffffffffffffff0ffffffff7da7fffffffffffffffffff0ffffd5ee71bffffffffffffffffffff0fffff47e67fffffffffffffffffffff0cffffe7c3ffafffffffffffffffffff03fffbec2e0fcfffffffffffffffffff0cffffff003affffffffffffffffffff0dfffffc00004fffffffffffffffffff0fff"`;

exports[`ImageLoader shall split the images properly 3`] = `"ffffffffffffffffffffc00000000000ffffffffffffffffffffc00000000000ffffffffffffffffffffe00000000000ffffffffffffffffffffe80000000000ffffffffffffffffffffc00000000000fffffffffffffffffffff00000000000ffffffffffffffffffffe00000000000ffffffffffffffffffffe00000000000fffffffffffffffffffff00000000000ffffffffffffff0007fff80000000000fffffffffffff8000000f80000000000ffffffffffffc0000000080000000000ffffffffffff00000000060000000000fffffffffffe00000000000000000000fffffffffff800400000000000000000fffffffffff006208000000000000000ffffffffffe00e620000000000000000ffffffffffc19eca0000000000000000ffffffffff81d4fe0800000000000000ffffffffff07e1fc0800000000000000fffffffffe0de3e82800000000000000fffffffffc1fc5f02080000000000000fffffffff83feff00300000000000000fffffffff87ffff00140000000000000fffffffff07ffff00240040000000000ffffffffe1fffff004e0000000000000fffffffff1fffff709c0000000000000ffffffffc3fffffe03a0000000000000ffffffff87fffffcd6a0000000000000ffffffff87fff7fd9d40008000000000ffffffff0fffe7ffab10008000000000ffffffff8fffc7cfaa30000000000000fffffffffffff3ffffffffe063ff337dffffffffffffe1ffffffeff021ff307bffffffffffffc0ffffffcff030ff906fffffffffffff80ffffffc738107f1858ffffffffffff807fffff8208187eb870ffffffffffff807fffff810008521a25ffffffffffff003fffff000408161020fffffffffffe001ffffe00040c323800fffffffffffe001ffffc000004320058fffffffffffe000ffffc000204191030fffffffffffc0007ffc0000006101800fffffffffffc0007ff80000002108850fffffffffff80007ff00000102040040fffffffffffc0003ff00000002000000fffffffffff80003fe00000003000000fffffffffff00001fe00000001000000fffffffffff00001fe00000081000000fffffffffff80001fe00000081800000fffffffffff80001fc00000081800000fffffffffff00001fc00000081800000fffffffffff00000fc000000c0800000fffffffffff80000fc00000040800000fffffffffff800007c000000c0800000fffffffffff000007c00000040800000fffffffffff000007c000000c0800000ffffffffffe000007800000040800000ffffffffffe000007800000040800000fffffffffff800007000000040800000fffffffffff0000060000000c0800000fffffffffff0000080000000c0800000ffffffffffe4000000000000e0000000ffffffffffe800000000000060800000fffffffffffc000000000000e3c001dffffffffffffc000000000000e3c0301ffffffffffffc000000000000e3c090fdfffffffffffc000000000000e3c18274fffffffffffc00006000000061c00038fffffffffff8000070000000e181006cfffffffffff8000060000000e3c00007fffffffffff80000f000000070800032fffffffffff80000f0000000f3800000fffffffffff80000f0000000e1800008fffffffffff80000f0000000e1800000fffffffffffc0000f8000000e1800000fffffffffffe000078000000e1800000fffffffffffe000078000280e1000000fffffffffffe000078000381e1000000fffffffffffe0000f8000341e1000000fffffffffffe0000fc0007d1c1000000fffffffffffe0001fc0007e1c0000000ffffffffffff0001fc000fe1c2000000ffffffffffff0003fe000fe1c2000000ffffffffffff0003ff000fe182000000fffffffdffff8003ff000fc386000000fffffffdffff8007ff800fc386000000fffffffeffff8007ff800f8704000000fffffffeffffc00fff800fe704000000fffffffeffffc00fffc00fce00000000fffffff67fffc01fffe01fce08000000fffffff77fffc01fffe01fce08000000ffffffff3fffc03fffe01f9c10000000fffffffb9fffe03fffe03f9c10000000fffffffb8fffe07ffff03e3800000000fffffffd87fff0fffff03e3800000000fffffffdcffffffffffffff000000000fffffffeefffffffffdffff040000000fffffffee7fffffffeb7fef040000000ffffffff71fffffffc87f5e080000000ffffffff71bfffffff03f1c000000000ffffffffb821ffffff87f18100000000ffffffffb820ffffff87c18000000000ffffffffdc78ffffffcfc30200000000ffffffffee7e57ffffcf820000000000ffffffffee1c07ffffff840000000000fffffffff70f03ffffff880000000000fffffffffb8f03ffffff100000000000fffffffff9c782ffffff200000000000fffffffffcc7f93ffffe400000000000fffffffffe61fa2ffffc800000000000ffffffffff30fc00fffb000000000000ffffffffff9c7f809ffe000000000000ffffffffffc63fe1bfd8000000000000ffffffffffe30fffff20000000000000fffffffffff083fff800000000000000fffffffffff800474000000000000000fffffffffff800000000000000000000fffffffffffc00000000000000000000fffffffffffe00000000000000000000ffffffffffff80000000000000000000ffffffffffffe0000000000000000000fffffffffffff80000000000000000007ffffffffffff8000000000000000000effffffffffffc0000000000000000006fffffffffffff100000000000000000063ffffffffffff800000000000000000033ffffffffffcc0000000000000000"`;

exports[`ImageLoader shall split the images properly 4`] = `"fbba4000001fffffffffffffffff0fff7fdf6000025ffbffffffffffffff0fff7eff600003ffffffffffffffffff0fff77e3e00000ffffffffffffffffff0ffffbdba00001ffffffffffffffffff0ffffedd40000bffffffffffffffffff0cfffff3600003ffffffffffffffffff0c7ffbefa00021ffffffffffffffffff0cfffbeee00017ff07ffffffffffffff00fffee6600013f8030007ffffffffff0cfffff0200007c00000007fffffffff0fffffdba0002f000000001fffffffff0ffffdf220002c0000000007ffffffff0fffffcae000380000000001ffffffff0ffffdd0e000e000000000007fffffff07fffeb1a000c000000000003fffffff0fff7f8d60000000000000001fffffff0ffff5cc200000000800000007ffffff0fffff88400000000040000003ffffff0ffffe29c00400020000002801ffffff0cfff46ec00000080400007800ffffff007ffef4c0000014040000f800ffffff0cffff9040000030000000f0007fffff0ffff9004040006000000160003fffff0fff880000000060000001a0001fffff0fffec48400000c0000002a0001fffff0fffb801400001a000000280000fffff07ffe80040000300000000c00007ffff0fffe80040000200000000800007ffff0fffe20240000600000008900003ffff0fff620040000500000005000003ffff0fffc000c0000400000000400001ffff047fffffdffc1fffffffffc0001dffff8060ffffdffc3fffffffff80001dffff8cc0ffffdff83fffffffff80000effff8fc8ffffdff87fffff7fff00000effff8fccffffdff87fa9b1bffe00000fffff8fccffffdff07f01102ffe0000077fff8fccffffdff0ff001017f2000003ffff87ccffffdfe0ff000004f0000003bfff8fccffffdfe1ff000000e8000007bfff8fccffffdfc1fe00000028000003ffff8fccffffdfc1fe00000010000001ffff8fccffffdfc7fe00000000000001ffff8cccffffdf87f800000000000001ffff8048ffffdf87f800000000000001ffff8cc0ffffdf83f800000000000005ffff8fe0ffffdf87f200000000000007ffff8fffffffdf87f800000000000007ffff8fffffffdf87ec00000000000007ffff8fffffff9f87e800000000000007ffff87ffffff9f07d800000000000007ffff87ffffff9f0fd800000000000003ffff87ffffff9f0fb800000000000003ffff87ffffff9f0fb000000000000003ffff87ffffff9f0f0000000000000003ffff806fffff9f0f0000000000000001ffff805fffffbf0f0000000000000001ffff807fffffbf0e00000000000000017fff87ffffffbe2e0000000000000000ffff87ffffffbe2f0000000000000000ffff87ffffffbe3f00000000000000007fff87ffffffbe3f80000000000000007fff87ffffffbe3f80000000000000003fff87ffffffbf3ff0000000000000003fff87ffffffbf3ff0000000000000003fff87ffffffbf3ff8000000000000003fff87ffffffbf3ff8000000000000003fff847fffffbf3ff800000020000000bfff807fffffbf3ff800000068000001ffff8461ffffbf3ff80000007000000fffff87c0ffffbf3ff80000007000001fffff87c8ffffbf3ff80000007000001fffff87ccffffff3ff80000007000000fffff87ccffffff3ffc0000007000000fffff87ccffffff1ffc0000002000001fffff87ccffffff1ffe0000000000000fffff87ccffff7f1ffe0000000000000fffff87ccffff7f1ffe0000000000000fffff87ccffff7f9ffe0000000000000fffff844cffff7f8fff0000000000000fffff804cffff7f8fff0000004000000fffff804cffff7f8fff000000c000000fffff8041ffff7f8fff000000e000000fffff8040ffff7fc7ff000000e0000015ffff8040ffff7fc7ff000000e0000009ffff8063ffff7fc7ff800000f0000009ffff80c7ffff7fc3ff800001f8000001ffff80c3ffff7fe3ff800003f8000003ffff80c3ffff7fe3ffc00003fc000003ffff80c3ffff7fe1ffc0000ffc000023bfff8043ffff7ff1ffc0000ffc000027ffff8003ffff7ff0ffc0000ffc0000477fff8003ffff7ff0ffc0000ffc000007ffff8003fffffff87fe0000ffe00000fffff8041fffffff87fe0000ffe00000effff8041fffffffc3ff0001ffe00000fffff8041fffffffc3ff0000ffe00001dffff8041fffffffe1ff8000ffe00000bffff8041fffffffe1ff8003fff00001bffff8041ffffffff0ffc003fff000017ffff8041fffeffff07fc007fff800037ffff8041fffeffff87fe007fff80002fffff8041fffeffff83fe00ffff80004fffff8001fffefffdc1fe00ffffc0005ff7ff8001fffeffffe1ff00ffffc0009fe7ff8001fffefffff0ff00ffffe0003f83ff0041fffeffff787fcbffffe0007f81ff0041fffeffff383ffffffff400ee01ff0040fffefffe1c1ffffffff000fa01ff0040fffeffff4e03fffffff001c000ff0040fffeffffb300ffffffe003c004ff0040fffefffff9807fffffc00740001f0000fffeffffe0c01fffff800000000f0000fffefffff00007fffe000000000f0000fffefffff80000fff800000000070001fffefffffe00000f0000000000060000fffeffffff0000000000000000020000fffeffffffc000000000000000020000fffeffffffd000000000000000000000fffefffffffc00000000000000000000fffefffffff330000000000000000000fffcfffffffc10000000000000000000fffcffffffec40000000000000000000fffc7ffffff600000000000000000000fffcffffffffe8800000000000000000fffc7fffffffc49c0000000000000000fffc7efe2c7a21e00000000000000000"`;
